## Exercises: Filter and query data

In [71]:
import pandas as pd
autos = pd.read_json("../Data/autos.json")
autos_column_names = autos.columns
autos_column_names

Index(['aspiration', 'body-style', 'bore', 'city-mpg', 'compression-ratio',
       'curb-weight', 'drive-wheels', 'engine-location', 'engine-size',
       'engine-type', 'fuel-system', 'fuel-type', 'height', 'highway-mpg',
       'horsepower', 'length', 'make', 'normalized-losses', 'num-of-cylinders',
       'num-of-doors', 'peak-rpm', 'price', 'stroke', 'symboling',
       'wheel-base', 'width'],
      dtype='object')

In [72]:
autos

aspiration   body-style  bore  city-mpg  compression-ratio  curb-weight  \
0          std  convertible  3.47        21                9.0         2548   
1          std  convertible  3.47        21                9.0         2548   
2          std    hatchback  2.68        19                9.0         2823   
3          std        sedan  3.19        24               10.0         2337   
4          std        sedan  3.19        18                8.0         2824   
..         ...          ...   ...       ...                ...          ...   
200        std        sedan  3.78        23                9.5         2952   
201      turbo        sedan  3.78        19                8.7         3049   
202        std        sedan  3.58        18                8.8         3012   
203      turbo        sedan  3.01        26               23.0         3217   
204      turbo        sedan  3.78        19                9.5         3062   

    drive-wheels engine-location  engine-size engine-type  ...         make  \
0            rwd           front          130        dohc  ...  alfa-romero   
1            rwd           front          130        dohc  ...  alfa-romero   
2            rwd           front          152        ohcv  ...  alfa-romero   
3            fwd           front          109         ohc  ...         audi   
4            4wd           front          136         ohc  ...         audi   
..           ...             ...          ...         ...  ...          ...   
200          rwd           front          141         ohc  ...        volvo   
201          rwd           front          141         ohc  ...        volvo   
202          rwd           front          173        ohcv  ...        volvo   
203          rwd           front          145         ohc  ...        volvo   
204          rwd           front          141         ohc  ...        volvo   

    normalized-losses  num-of-cylinders  num-of-doors  peak-rpm    price  \
0                 NaN              four           two    5000.0  13495.0   
1                 NaN              four           two    5000.0  16500.0   
2                 NaN               six           two    5000.0  16500.0   
3               164.0              four          four    5500.0  13950.0   
4               164.0              five          four    5500.0  17450.0   
..                ...               ...           ...       ...      ...   
200              95.0              four          four    5400.0  16845.0   
201              95.0              four          four    5300.0  19045.0   
202              95.0               six          four    5500.0  21485.0   
203              95.0               six          four    4800.0  22470.0   
204              95.0              four          four    5400.0  22625.0   

    stroke  symboling wheel-base width  
0     2.68          3       88.6  64.1  
1     2.68          3       88.6  64.1  
2     3.47          1       94.5  65.5  
3     3.40          2       99.8  66.2  
4     3.40          2       99.4  66.4  
..     ...        ...        ...   ...  
200   3.15         -1      109.1  68.9  
201   3.15         -1      109.1  68.8  
202   2.87         -1      109.1  68.9  
203   3.40         -1      109.1  68.9  
204   3.15         -1      109.1  68.9  

[205 rows x 26 columns]

**Use the "autos" dataset to answer the following questions:**

**Q1:** How many cars from Jaguar are in the dataset, and what is the price of the most expensive one?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;count=3
&nbsp;&nbsp;&nbsp;price=36000.0
</details>

In [73]:
autos.dropna(subset=["price"], inplace=True)
autos.sort_values(by="price", ascending=False, inplace=True)

number_of_jaguars = len(autos.query("make == 'jaguar'"))
most_expensive_jaguar_price = autos[autos['make'] == 'jaguar']['price'].iloc[0]

print(f"Number of Jaguar cars: {number_of_jaguars}")
print(f"Price of the most expensive Jaguar: {most_expensive_jaguar_price}")

Number of Jaguar cars: 3
Price of the most expensive Jaguar: 36000.0


**Q2:** How many cars from Toyota are in the dataset, and what is the price of the most expensive one?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;count=32
&nbsp;&nbsp;&nbsp;price=17669.0
</details>

In [74]:
number_of_toyotas = len(autos.query("make == 'toyota'"))
most_expensive_toyota_price = autos[autos['make'] == 'toyota']['price'].iloc[0]

print(f"Number of Toyota cars: {number_of_toyotas}")
print(f"Price of the most expensive Toyota: {most_expensive_toyota_price}")

Number of Toyota cars: 32
Price of the most expensive Toyota: 17669.0


**Q3:** What is the length, width and height of the most expensive car in the entire dataset?

<details>
<summary>Answer</summary>
<br>&nbsp;&nbsp;&nbsp;length=199.2
&nbsp;&nbsp;&nbsp;width=72.0
&nbsp;&nbsp;&nbsp;height=55.4
</details>

In [75]:
most_expensive_car = autos.iloc[0]

most_expensive_car_make = most_expensive_car['make']
most_expensive_car_price = most_expensive_car['price']
most_expensive_car_length = most_expensive_car['length']
most_expensive_car_width = most_expensive_car['width']
most_expensive_car_height = most_expensive_car['height']

print(f"The most expensive car is a {most_expensive_car_make} and costs {most_expensive_car_price}. It's length is {most_expensive_car_length}, it's width {most_expensive_car_width}, and it's height {most_expensive_car_height}.")

The most expensive car is a mercedes-benz and costs 45400.0. It's length is 199.2, it's width 72.0, and it's height 55.4.


**Q4:** What is the lowest price per horsepower in the dataset, and what brand ("make") is that car?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;price per horsepower=72.84
&nbsp;&nbsp;&nbsp;brand=Toyota
</details>

In [76]:
autos['price per hp'] = autos['price'] / autos['horsepower']
autos.dropna(subset=["price per hp"], inplace=True)
autos.sort_values(by="price per hp", ascending=True, inplace=True)

lowest_price_per_hp_make = autos['make'].iloc[0]
lowest_price_per_hp = autos['price per hp'].iloc[0]

print(f"The car with the lowest price per horsepower is a {lowest_price_per_hp_make} and it's price per hp is {lowest_price_per_hp}.")

The car with the lowest price per horsepower is a toyota and it's price per hp is 72.83620689655173.


**Q5:** How many of the cars in the dataset has as many cylinders as they have doors?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;cars=95
</details>

In [77]:
autos.dropna(subset=["num-of-cylinders", "num-of-doors"], inplace=True)
cars_same_cylinders_doors = len(autos.query("`num-of-cylinders` == `num-of-doors`"))
print(cars_same_cylinders_doors)

93


**Use the "autos" dataset and write python code to solve the following tasks:**

**T1:** Print a string to inform the user of the price difference between the cheapest and the most expensive car in the dataset.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example:</b><br>
&nbsp;&nbsp;&nbsp;The price difference between the cheapest and the most expensive car is 40282.0
</details>

In [78]:
autos.sort_values(by="price", ascending=True, inplace=True)

highest_price = autos['price'].iloc[-1]
lowest_price = autos['price'].iloc[0]

print(f"The difference between the most expensive and the cheapest car is: {highest_price - lowest_price}")

The difference between the most expensive and the cheapest car is: 40282.0


**T2:** Ask the user to input a brand, then print the price range for that brand.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 1:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: volvo<br>
&nbsp;&nbsp;&nbsp;The prices for cars of brand 'volvo' ranges from 12940.0 to 22625.0<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 2:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: toyota<br>
&nbsp;&nbsp;&nbsp;The prices for cars of brand 'toyota' ranges from 5348.0 to 17669.0<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 3:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: tesla<br>
&nbsp;&nbsp;&nbsp;The brand 'tesla' does not exists in the dataset.<br>
</details>

In [81]:
car_make = input("Please enter the make of the car:")
cars_input_make = autos.query('make == @car_make')

highest_price = cars_input_make['price'].iloc[-1]
lowest_price = cars_input_make['price'].iloc[0]

price_difference = highest_price - lowest_price

print(price_difference)

9685.0


**T3:** Ask the user to input a brand, then print the number of cars in the dataset for that brand, and all attributes for a random sample car of that brand.
<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: mazda<br>
&nbsp;&nbsp;&nbsp;There are 17 cars of brand 'mazda' in the dataset.<br><br>
&nbsp;&nbsp;&nbsp;Here is the data for a random 'mazda' car:<br>
&nbsp;&nbsp;&nbsp;aspiration = std<br>
&nbsp;&nbsp;&nbsp;body-style = sedan<br>
&nbsp;&nbsp;&nbsp;bore = 3.03<br>
&nbsp;&nbsp;&nbsp;city-mpg = 31<br>
&nbsp;&nbsp;&nbsp;compression-ratio = 9.0<br>
&nbsp;&nbsp;&nbsp;curb-weight = 1945<br>
&nbsp;&nbsp;&nbsp;drive-wheels = fwd<br>
&nbsp;&nbsp;&nbsp;engine-location = front<br>
&nbsp;&nbsp;&nbsp;engine-size = 91<br>
&nbsp;&nbsp;&nbsp;engine-type = ohc<br>
&nbsp;&nbsp;&nbsp;fuel-system = 2bbl<br>
&nbsp;&nbsp;&nbsp;fuel-type = gas<br>
&nbsp;&nbsp;&nbsp;height = 54.1<br>
&nbsp;&nbsp;&nbsp;highway-mpg = 38<br>
&nbsp;&nbsp;&nbsp;horsepower = 68.0<br>
&nbsp;&nbsp;&nbsp;length = 166.8<br>
&nbsp;&nbsp;&nbsp;make = mazda<br>
&nbsp;&nbsp;&nbsp;normalized-losses = 113.0<br>
&nbsp;&nbsp;&nbsp;num-of-cylinders = four<br>
&nbsp;&nbsp;&nbsp;num-of-doors = four<br>
&nbsp;&nbsp;&nbsp;peak-rpm = 5000.0<br>
&nbsp;&nbsp;&nbsp;price = 6695.0<br>
&nbsp;&nbsp;&nbsp;stroke = 3.15<br>
&nbsp;&nbsp;&nbsp;symboling = 1<br>
&nbsp;&nbsp;&nbsp;wheel-base = 93.1<br>
&nbsp;&nbsp;&nbsp;width = 64.2<br>
</details>

In [80]:
car_make_input = input("Please enter the make of the car:")
cars_input_make = autos.query('make == @car_make_input')


number_of_input_cars = len(cars_input_make)

print(f"Input the name of a brand: {car_make_input}")
print(f"There are {number_of_input_cars} cars of brand '{car_make_input}' in the dataset.")

print()
print(f"Here is the data for a random '{car_make_input}' car:")

sample_car = cars_input_make.sample(1).iloc[0]

for attribute, value in sample_car.items():
    print(f"{attribute} = {value}")

Input the name of a brand: volvo
There are 11 cars of brand 'volvo' in the dataset.

Here is the data for a random 'volvo' car:
aspiration = std
body-style = wagon
bore = 3.7800000000000002
city-mpg = 24
compression-ratio = 9.5
curb-weight = 3042
drive-wheels = rwd
engine-location = front
engine-size = 141
engine-type = ohc
fuel-system = mpfi
fuel-type = gas
height = 57.5
highway-mpg = 28
horsepower = 114.0
length = 188.8
make = volvo
normalized-losses = 74.0
num-of-cylinders = four
num-of-doors = four
peak-rpm = 5400.0
price = 16515.0
stroke = 3.15
symboling = -1
wheel-base = 104.3
width = 67.2
price per hp = 144.8684210526316


**T4:** Ask the user to input a brand, then export all cars of that brand to a .csv file with the same name as the brand.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 1:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: volkswagen<br>
&nbsp;&nbsp;&nbsp;Exported 12 cars to 'volkswagen.csv'<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 2:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: tesla<br>
&nbsp;&nbsp;&nbsp;The brand 'tesla' does not exists in the dataset.<br>
</details>